### PR TITLE
introduce lazy loading for resources

### PR DESCRIPTION
**Pull request by labuzm**

### CODE REVIEW

File: include/kaacore/resources.h [hunk 1]
```diff
@@ -1,19 +1,77 @@
...
+        return ptr.get();
+    }
+};
+
+template<typename Key_T, typename Resource_T>
+class ResourcesRegistry {
+  public:
+    void initialze()
+    {
+        for (auto it = this->_registry.begin(); it != this->_registry.end();
```
Comment (maniek2332) [inline include/kaacore/resources.h]: can be simplified as `for (auto& [key, value] : this->_registry`

File: include/kaacore/resources.h [hunk 2]
```diff
@@ -1,19 +1,77 @@
...
+             ++it) {
+            it->second.lock()->_initialize();
+        }
+    }
+    void uninitialze()
+    {
+        for (auto it = this->_registry.begin(); it != this->_registry.end();
```
Comment (maniek2332) [inline include/kaacore/resources.h]: can be simplified as `for (auto& [key, value] : this->_registry`

File: include/kaacore/resources.h [hunk 3]
```diff
@@ -1,19 +1,77 @@
...
+};
+
+template<typename Key_T, typename Resource_T>
+class ResourcesRegistry {
+  public:
+    void initialze()
+    {
+        for (auto it = this->_registry.begin(); it != this->_registry.end();
+             ++it) {
+            it->second.lock()->_initialize();
```
Comment (maniek2332) [inline include/kaacore/resources.h]: unhandled case when `std::weak_ptr::lock` returns empty `std::shared_ptr`

File: include/kaacore/resources.h [hunk 4]
```diff
@@ -1,19 +1,77 @@
...
+        }
+    }
+    void uninitialze()
+    {
+        for (auto it = this->_registry.begin(); it != this->_registry.end();
+             ++it) {
+            it->second.lock()->_uninitialize();
```
Comment (maniek2332) [inline include/kaacore/resources.h]: unhandled case when `std::weak_ptr::lock` returns empty `std::shared_ptr`

File: include/kaacore/resources.h [hunk 5]
```diff
@@ -1,19 +1,77 @@
...
+        }
+    }
+    void register_resource(
+        const Key_T& key, const std::weak_ptr<Resource_T> resource)
+    {
+        KAACORE_ASSERT(this->_registry.find(key) == _registry.end());
```
Comment (maniek2332) [inline include/kaacore/resources.h:57]: this should allow registering resource if currently held `std::weak_ptr` is already expired

File: src/resources.cpp [hunk 6]
```diff
@@ -1,3 +1,7 @@
 #include "kaacore/resources.h"
 
-namespace kaacore {} // namespace kaacore
+namespace kaacore {
+
+Resource::~Resource() {}
```
Comment (maniek2332) [inline src/resources.cpp:5]: can we put:
```
    if (this->is_initialized) {
        this->_uninitialize();
    }
```
here, co it does not repeat in desended destructors?

**Review by maniek2332 — CHANGES_REQUESTED** [review]
see compiler notes:
```
../src/images.cpp:101:15: warning: using the result of an assignment as a condition without parentheses [-Wparentheses]
    if (image = _images_registry.get_resource(path)) {
        ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/images.cpp:101:15: note: place parentheses around the assignment to silence this warning
    if (image = _images_registry.get_resource(path)) {
              ^
        (                                          )
../src/images.cpp:101:15: note: use '==' to turn this assignment into an equality comparison
    if (image = _images_registry.get_resource(path)) {
              ^
              ==
../src/audio.cpp:50:20: warning: using the result of an assignment as a condition without parentheses [-Wparentheses]
    if (sound_data = _sound_registry.get_resource(path)) {
        ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/audio.cpp:50:20: note: place parentheses around the assignment to silence this warning
    if (sound_data = _sound_registry.get_resource(path)) {
                   ^
        (                                              )
../src/audio.cpp:50:20: note: use '==' to turn this assignment into an equality comparison
    if (sound_data = _sound_registry.get_resource(path)) {
                   ^
                   ==
../src/audio.cpp:139:20: warning: using the result of an assignment as a condition without parentheses [-Wparentheses]
    if (music_data = _music_registry.get_resource(path)) {
        ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/audio.cpp:139:20: note: place parentheses around the assignment to silence this warning
    if (music_data = _music_registry.get_resource(path)) {
                   ^
        (                                              )
../src/audio.cpp:139:20: note: use '==' to turn this assignment into an equality comparison
    if (music_data = _music_registry.get_resource(path)) {
                   ^
                   ==
../src/fonts.cpp:192:19: warning: using the result of an assignment as a condition without parentheses [-Wparentheses]
    if (font_data = _fonts_registry.get_resource(path)) {
        ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/fonts.cpp:192:19: note: place parentheses around the assignment to silence this warning
    if (font_data = _fonts_registry.get_resource(path)) {
                  ^
        (                                             )
../src/fonts.cpp:192:19: note: use '==' to turn this assignment into an equality comparison
    if (font_data = _fonts_registry.get_resource(path)) {
                  ^
                  ==
```

File: src/resources.cpp (hunk 6 above)
Comment (labuzm) [reply src/resources.cpp]: nope, you are now allowed to call pure virtual function from destructor

File: include/kaacore/resources.h (hunk 5 above)
Comment (labuzm) [reply include/kaacore/resources.h]: This won't happen since registry entry is erased on resource deletion

**Review by maniek2332 — APPROVED** [review]

**Pull request merged**